### ISSUE TITLE
For SDK runtime stability guarantees do *not* use Generics upon Go 1.18 release

### ISSUE THREAD

**Issue by berndverst** [Azure/azure-sdk-for-go#16195]
**Generics are scheduled to be released with Go 1.18. This feature may not be ready for production use according to the Golang team.** As such we should take a **conservative approach** and **hold-off on the use of Generics** with the release of Go 1.18.

I propose not to use Generics until later minor (not patch) Go language releases, (e.g. 1.20+)

> Because we will not know what the best practices are for using generics, our docs will not be able to give precise, definitive answers about when to use them and when not to. We can and will still give rough guidelines. For comparison, we wrote the initial version of Effective Go after an intense full year of writing Go code non-stop. We don't have the same level of experience yet with generics, so while we will of course provide docs about how to use generics, we can't provide anything as prescriptive about style and best practices. We simply don't know them yet.
> [...]
> 
> Because we will not have any production experience with generics, we will make clear in the release notes that production uses of generics should be approached with appropriate caution. This is not a criticism of the team’s excellent work. It is simply an observation that generics is different from most Go changes. When we rewrite the garbage collector or change the calling convention, we run all of Google’s Go programs, in test and in production, using the new implementation, which exercises the change quite a bit and shakes out hard-to-find bugs. In contrast, rebuilding non-generic code with the in-progress Go 1.18 toolchain does not exercise the support for generics, meaning we can't build the same confidence.

https://groups.google.com/g/golang-dev/c/iuB22_G9Kbo

To be clear, the concern here is *not* the fact that the use of Generics is a breaking change forcing downstream users to use Go 1.18.

**The concern is around the runtime stability of this language feature and the impact it could have on downstream consumers.**

**Comment by berndverst** [comment]
FYI I work on Dapr

**Comment by berndverst** [comment]
cc @greenie-msft @paulyuk

**Comment by berndverst** [comment]
Dapr makes guarantees about component (integration) stability and runtime performance. We cannot migrate to the new "Track 2" SDKs if they were to use Go language features that aren't guaranteed to be stable.

If the use of Generics were to introduce problems - either in memory consumption, latency or unexpected errors (especially the kind not observable in unit tests) neither Dapr nor the Azure SDK team would not be able to remedy these problems in an emergency hotfix release.

**Comment by JeffreyRichter** [comment]
I have been tracking Google's Go team and the compiler/language/tools very closely in regards to many things including Generics. They have very high confidence in the generic language syntax and the compiler support. They have lower confidence in producing new standard library generic functions/types due to the lack of experience you site.

The Azure SDK has very few constructs: 
- We define Client types (these will never be generic)
- The methods on these types take parameters (these will never be generic).
- The methods return a response (we do want to have a generic Response structure type) and this should be perfectly fine.
- Some methods return a Poller or a Pager response and we also want these to be generic and this should also be fine.

If we ever use Go's libraries that offer generic types/functions these will be part of our implementation and never exposed to a customer anyway. We actually do not expect to need any of this functionality. But, if we did, we could write it ourself and change it later, or we could copy the code from a Go experimental package into our package and then change it later, or we could take a dependency on a Go experimental package and then change this dependency later to our own code or to code in a Go standard package when it becomes available in the future. 

In other words, I do not foresee adopting generic types early in our API surface area as being a problem for the Azure Go SDK in any way.
